File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    n = len(board)
    stack = [[] for _ in range(n)]
    
    for i in range(n-1, -1, -1):
        line = board[i]
        for i, num in enumerate(line):
            if num == 0:
                continue
            stack[i].append(num)
    
    basket = []
    answer = 0
    
    for move in moves:
        cur_idx = move - 1
        if not stack[cur_idx]:
            continue
        elif basket and stack[cur_idx][-1] == basket[-1]:
            stack[cur_idx].pop()
            basket.pop()
            answer += 2
        else:
            basket.append(stack[cur_idx].pop())
    
    
    return answer

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int N = board.length;
        
        Stack<Integer>[] lines = new Stack[N + 1];
        Stack<Integer> basket = new Stack<>();
        
        int answer = 0;
        
        for(int i = 0; i < N; i++){
            lines[i + 1] = new Stack<>();
            for(int j = N - 1; j >= 0; j--){                
                if(board[j][i] == 0) break;
                lines[i + 1].add(board[j][i]);
            }            
        }
        
        for(int move:moves){
            if(lines[move].isEmpty()) continue;
            int curNum = lines[move].pop();
            
            if(!basket.isEmpty() && basket.peek() == curNum){
                answer += 2;
                basket.pop();
            } else{
                basket.push(curNum);
            }
        }
        
        return answer;
    }
}